examples/crowd-funding/tests/campaign_lifecycle.rs

```diff
@@ -0,0 +1,214 @@
+                        amount: pledge_amount,
+                    },
+                );
+            })
+            .await;
+
+        assert_eq!(pledge_effects.len(), 3);
+        pledges_and_transfers.extend(pledge_effects);
+    }
+
+    campaign_chain
+        .add_block(|block| {
+            block.with_incoming_messages(pledges_and_transfers);
+        })
+        .await;
+
+    assert_eq!(
+        FungibleTokenAbi::query_account(token_id, &campaign_chain, campaign_account).await,
+        None
+    );
+
+    campaign_chain
+        .add_block(|block| {
+            block.with_operation(campaign_id, Operation::Collect);
+        })
+        .await;
+
+    assert_eq!(
+        FungibleTokenAbi::query_account(token_id, &campaign_chain, campaign_account).await,
+        Some(
+            pledge_amount
+                .saturating_add(pledge_amount)
+                .saturating_add(pledge_amount)
+        ),
+    );
+
+    for (backer_chain, backer_account, initial_amount) in backers {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id, &backer_chain, backer_account).await,
+            Some(initial_amount.saturating_sub(pledge_amount)),
+        );
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id, &campaign_chain, backer_account).await,
+            Some(Amount::from(0)),
+        );
+    }
+}
+
+/// Test creating a campaign and cancelling it.
+///
+/// Creates a campaign on a `campaign_chain` and sets up the fungible token to use with three
+/// backer chains. Pledges part of each backer's balance to the campaign and then completes it,
+/// collecting the pledges. The final balance of each backer and the campaign owner is checked.
+#[tokio::test]
+async fn cancel_successful_campaign() {
+    let initial_amount = Amount::from(100);
+    let target_amount = Amount::from(220);
+    let pledge_amount = Amount::from(75);
+
+    let (validator, bytecode_id) = TestValidator::with_current_bytecode().await;
+
+    let fungible_publisher_chain = validator.new_chain().await;
+    let mut campaign_chain = validator.new_chain().await;
+    let campaign_account = AccountOwner::from(campaign_chain.public_key());
+
+    let fungible_bytecode_id = fungible_publisher_chain
+        .publish_bytecodes_in("../fungible")
+        .await;
+
+    let (token_id, backers) = FungibleTokenAbi::create_with_accounts(
+        &validator,
+        fungible_bytecode_id,
+        iter::repeat(initial_amount).take(3),
+    )
+    .await;
+
+    let campaign_state = InitializationArgument {
+        owner: campaign_account,
+        deadline: Timestamp::from(10),
+        target: target_amount,
+    };
+    let campaign_id = campaign_chain
+        .create_application::<CrowdFundingAbi>(
+            bytecode_id,
+            token_id,
+            campaign_state,
+            vec![token_id.forget_abi()],
+        )
+        .await;
+
+    let mut pledges_and_transfers = Vec::with_capacity(2 * backers.len());
```

Comment:
3?

…or just `Vec::new()`? I don't think this is what makes our tests slow. :smile: